NAAM.md

```diff
@@ -0,0 +1,278 @@
+see [Updating IPNS Records](#updating-ipns-records).
+
+Note that the IPNS multicodec code `0xe5` is tagged as "namespace" in the CSV table, which could
+signify an IPNS path, not record. There is currently no multicodec code for IPNS records themselves.
+Alternatively, a new multicodec code can be introduced to the
+[multicodec CSV file](https://github.com/multiformats/multicodec/blob/master/table.csv) to avoid
+potential confusion between the two. The new code can then be used as the Protocol ID of Naam
+Metadata.
+
+### Updating IPNS Records
+
+IPNI offers a built-in mechanism to efficiently update the metadata associated to a list of
+multihashes without having to republish the entries. Indexer nodes uniquely identify the metadata by
+the advertisement `Provider`, and `ContextID`.
+
+Both of these values remain the same in Naam advertisements: the former is the same peer ID that
+forms part of the IPNS key, i.e. `/ipns/<peer-id>`, and the latter is fixed. As a result, to update
+an IPNS record, all that's needed is to publish a new Naam advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `false`, and
+* has the updated IPNS record in Naam Metadata format.
+
+Note that there is no need to specify `Entries` when updating an existing record. It is inferred by
+the network indexers since `Provider` and `ContextID` in Naam advertisements never change.
+
+### Removing IPNS Records
+
+IPNS has EOL and TTL features built-in. This means explicit removal of records is not strictly
+required. Instead, records eventually expire. Though not strictly necessary, Naam protocol can
+utilise IPNI to explicitly remove records just like a traditional DNS nameserver.
+
+The process to remove an IPNS record published via Naam protocol is identical to removal
+advertisement specified by IPNI. Similar to a regular advertisement, an IPNS record may be removed
+by publishing an advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `true`, and
+* sets `Entries` link to `NoEntries` as specified by IPNI.
+
+Once ingested, indexer nodes will eventually remove the Naam advertisement metadata along with the
+publisher information.
+
+### Resolving IPNS Records
+
+IPNS record resolution involves finding the IPNS record associated to a given IPNS key, i.e.
+`/ipns/<peer-id>`. As eluded to earlier, Naam uses the find API already provided by the network
+indexers to achieve this. To resolve an IPNS record, a resolver:
+
+* calculates the multihash of the given IPNS key as its SHA-256,
+* looks up the provider info for that multihash via the IPNI find API,
+* decodes the metadata from the provider info as IPNS record,
+* validates the record, and
+* if valid returns its value.
```

Comment:
> there is at most 1 record worth returning
Yep; the go implementation does exactly that, in that in the context of naming over advertisement, only the "head" advertisement matters, and because advertisements are signed we can verify that the advertisement containing _the_ IPNS record was indeed published by the peer that corresponds to the IPNS key.

> anyone can store and therefore store with IPNI an IPNS record signed by Alice
Yes I think of this as a good thing, since even if Alice is offline, IPNS key is resolvable. I think the same is true about the existing DHT-based IPNS resolution?

> decode all of the returned values and calculate the best themselves...
> if the IPNI nodes do validation before storing the data in their database
Two things to point out here: 
* Because the provider and context ID is identical across all NAAM advertisements, there will _only_ be a single record as a result of processing those ads.
* The IPNI built-in validation, checks the signature of the ad and would guarantee that the ad is from the publisher/provider. 

> do the work of selecting the "best" record in between the IPNI database and the HTTP API
Thank you for mentioning the HTTP API, that seems reasonable, though I must point out in IPNI there will only be one IPNS record for a given IPNS key. Reading this comment I am worried that I have not spelled out the fact that only the head advertisement matters, and because provider and context ID are identical across all NAAM advertisements then there will only be a single provider record for it and therefore a singe IPNS record. 

Having pointed out the singularity of IPNS records resolved via IPNI,  selecting the best between HTTP and API, would become a matter of EOL value and whichever record is newer I _think_.

Reply:
>  and because advertisements are signed we can verify that the advertisement containing the IPNS record was indeed published by the peer that corresponds to the IPNS key.

You can't really do this. There may be many IPNS keys owned by a single peer, Alice. Of course she can also use her libp2p node ID as a key but this isn't the only one.

> though I must point out in IPNI there will only be one IPNS record for a given IPNS key.

I'm not sure if I'm misunderstanding or misexplaining myself. Say the following happens:
1. On day 1 Alice signs an IPNS record R1 using key K that's valid for 30 days and has a sequence number 1, then advertises it to IPNI from the AliceProvider
2. On day 2 Bob sees Alice's IPNS record R1 and says "hey, I should try and keep that alive" so advertises it to IPNI from the BobProvider
3. On day 3 Alice signs an IPNS record R2 using key K that's valid for 30 days and has a sequence number 2, then advertises it to IPNI from the AliceProvider
4. On day 4 when Charlie tries to look up the record associated with the IPNS identifier derived from K from IPNI there are two records (AliceProvider, R2) and (BobProvider, R1) which will either both need to be returned (for the client to figure out which is valid) or the IPNI node will need to figure out that R2 > R1 and only publish the result from the AliceProvider
    - Note: In theory R2 could be invalid (EOL was on day -7, or even day 3.5, invalid signature, etc.) so just choosing R2 because Alice published R1 and R2 to IPNI isn't sufficient

Reply:
On step `2`: this is where I need your help to get my head around this; I suspect I am missing something fundamental about IPNS. 
In Naam, `BobProvider` can only advertise IPNS records with key `/ipns/<bob's-key>`, because Bob does not have Alice's key. What am I missing?

On step `4`: thank you for pointing out the case where R2 could have expired or invalid. As the spec stands, such publication by Alice would make R1 no longer discoverable. I am curious why should R1 remain discoverable if Alice has published a new record for the same key that has expired? One could argue because the key is the same and the value is changed, then the value corresponding to the key should be the latest, i.e. the expired record and it is Alice's responsibility to publish new records to correct it. As for the case, where signature of R2 is invalid, again one can argue that it is Alice's responsibility to publish correct IPNS records if Alice wants the records to be discoverable.  

WDYT?

Reply:
> On step 2: this is where I need your help to get my head around this; I suspect I am missing something fundamental about IPNS.

> In Naam, BobProvider can only advertise IPNS records with key /ipns/<bob's-key>, because Bob does not have Alice's key. What am I missing?

I thought I was explaining this above, but I'll try to give more examples to see if that helps.

Alice and Bob can each have hundreds of keys (e.g. Ed25519 key pairs) for a variety of purposes. Any of these keys can be used to create IPNS identifiers, such as `/ipns/alices-favorite-book-key`, `/ipns/alices-least-favorite-movie-key`,  etc. These may/may not be the same as the `alices-ipni-provider-node-key` or `alices-laptop-kubo-node-libp2p-key`.

This means that this approach (i.e. one key per IPNI provider) is not sufficient, nor necessary https://github.com/ipni/go-naam/blob/9002ac91a0e3d606d81748df34d63e5f826e821a/naam.go#L142-L143.

Additionally, once Alice has signed a record R1 for `/ipns/alices-favorite-book-key` with her private key there's nothing stopping anyone else (e.g. Bob) from publishing R1 as well since that signed record is now public information.

> On step 4: thank you for pointing out the case where R2 could have expired or invalid. As the spec stands, such publication by Alice would make R1 no longer discoverable. I am curious why should R1 remain discoverable if Alice has published a new record for the same key that has expired?

Given the above let's revisit. Let's take the go-naam implementation as-is as an example. We can see that there is no server side code at all, only client side code. This means a few things:
1. IPNS records are not validated server side, meaning they might be completely invalid (missing a signature, signed with the wrong key, etc.) not just expired
2. Anyone can publish R1 including both Alice and Bob AND there is no way for IPNI to know whether Alice or Bob is the "canonical" author.
3. 1 + 2 implies that only sending back a single value would be unsafe since there's no logic to decide which one to send back. This is fixed if there's server-side logic deciding which to send back (i.e. negating 1).

This means that as is https://github.com/ipni/go-naam/blob/9002ac91a0e3d606d81748df34d63e5f826e821a/indexer_client.go#L105-L110 seems to be a DoS vulnerability since a malicious Mallory could advertise `/ipns/alice-favorite-book-key` with a totally bogus and invalid value (i.e. not signed). This is because a) the server doesn't validate the publish b) the server sends back all matches for a given multihash and c) the client code chooses the first response it might be Mallory's which would prevent resolving the IPNS address.

The solutions here are, as mentioned above, either:
1. Have the client process all the results and choose the best valid one (the server will already send back multiple results as-is). go-ipns has tools for this.
2. Have the server do processing and only store the best valid entry and return only that one

@masih does that make sense?

Reply:
It does thank you for helping me understand. Would [this](https://github.com/ipni/go-naam/commit/1a9f27b7d8bbab3340a0a11f02a2fcdab4f17c34) address your concern? This is to make it more explicit that the identity used to sign and publish the advertisement must match the identity in IPNS key. 

This specification is written to require zero changes on the existing ad ingest pipeline in the hope to make it easier to integrate. Relaxing that condition slightly, it should be straightforward to also expand server-side validations.

Reply:
I'm not sure. IIRC `pr.Provider` from [here](https://github.com/ipni/go-naam/commit/1a9f27b7d8bbab3340a0a11f02a2fcdab4f17c34#diff-170f0ccd11e315be059827c8f1ee6758cc9cb83a91802344497b8f47a6bfb2b1R107) refers to an the "provider" of a CID not the index provider node and there's no proven relationship the owner of `AliceIndexProviderKey` is the owner of `AliceMHProviderKey` (e.g. for example, I could delegate all my IPNI record publishes to a third party). IIUC the plan is to have some reputation to track good behavior here, but it's TBD and best-effort.

However, even if this did work it would limit utility by only allowing 1 key per index provider (e.g. really bad for infra providers, or just anyone with multiple mutable websites to manage under different keys). Given that your proposed change is client side, why not just use go-ipns to [select](https://github.com/ipfs/go-ipns/blob/1af8e5c36767c858a02d1d4bab6357199d62e906/record.go#L90) the best record?

For example in go-naam do something roughly like:

```
ipnsValidator := ipns.Validator{}
var records [][]byte
for _, pr := range mhr.ProviderResults {
  if !bytes.Equal(pr.ContextID, ContextID) { continue }
  recBytes, err := getIPNSBytesFromMetadata(pr.Metadata)
  if err != nil { continue }
  if ipnsValidator.Validate(ipnsID, pr.Metadata) != nil { continue }
  records = append(records, recBytes)
}
bestRecordIndex, err := ipnsValidator.Select(ipnsID, records)
if err != nil || bestRecordIndex <0 { return nil, routing.ErrNotFound}
return records[bestRecordIndex], nil
```

Basically, because the validation is only happening client-side anyway why not do a little more CPU work (rather than relying on assumptions) in the validation logic in order to unlock a lot of functionality. If it becomes important later servers can add the validation onto their end.

Reply:
> proven relationship the owner of AliceIndexProviderKey is the owner of AliceMHProviderKey

Advertisement [signature](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L48) would prove that which is [verified](https://github.com/ipni/storetheindex/blob/main/internal/ingest/linksystem.go#L92) at the time of ingest by indexer nodes. Right?

> only allowing 1 key per index provider

That's a reasonable concern. Though, I must point out, Naam publish is way more light-weight than a regular [index provider](https://github.com/ipni/index-provider) set up. The two also are not bound to each other: a user with multiple identities can publish different advertisement chains under different identities. The Naam ad chains can be published on their own exclusive chain.

> why not do a little more CPU work

go-naam already does the validation [here](https://github.com/ipni/go-naam/blob/main/naam.go#L116). Is there something I have missed in validating the records?

Thank you for pointing out `select` I missed it! I still think using `select` doesn't make sense because, indexer has to return a single record for the same Provider ID + Context ID + Multihash. If it does not then there is bug in the indexer implementation. When the result can only be a single one then `select` usage doesn't quite make sense.

Reply:
> Advertisement [signature](https://github.com/ipni/storetheindex/blob/main/api/v0/ingest/schema/schema.ipldsch#L48) would prove that which is [verified](https://github.com/ipni/storetheindex/blob/main/internal/ingest/linksystem.go#L92) at the time of ingest by indexer nodes. Right?

Ah ok thx. I knew you could publish for others, but I guess that depends on the policy. Not sure how that's determined, but if it's sufficiently trusted then that's probably ok.

> The Naam ad chains can be published on their own exclusive chain.

For sure, and maybe this depends on what the delegated advertising policies look like but it'd be annoying to have to have hundreds of libp2p hosts (and ports) for hundreds of naam-index-provider nodes. If it's as simple as making all the private keys available to the a single libp2p index provider node that's probably ok (although note there might be some operational complexity around creating advertisements if people want to isolate their private keys to colder storage/hardware devices).

> I still think using select doesn't make sense because, indexer has to return at most one record for the same Provider ID + Context ID + Multihash.

Right, this is for enabling multiple return values. The only use case I can think of for this at the moment is to increase some resiliency of the IPNS record which is not currently allowed. For example:
1) Alice decides to make a [consistency vs availability tradeoff](https://docs.ipfs.tech/concepts/ipns/#tradeoffs-between-consistency-vs-availability) and always publish her records with EOLs of effectively infinity
2) At some point Alice takes her node (and index provider) offline
3) A few days later index providers garbage collect her records
4) Bob tries to keep Alice's record alive by publishing it to IPNI, but can't which means `/ipns/alice-key` is effectively dead even though it doesn't have to be

This isn't necessarily the most important, although it's something that's being lost here. It could of course be added later as well if needed.

Reply:
> If it's as simple as making all the private keys available to the a single libp2p index provider node that's probably ok

yup that should suffice. 

>  it's something that's being lost here.

Agreed; already added as limitations. As is if the provider side is offline for longer than a week records are no longer resolvable. Server side changes to treat Naam context ID differently can be a way rectify this.

Thank you so much for the great pointers.